File: Plugins/AM_Decon2ls_PlugIn_Decon2LSV2/clsDeconToolsQCPlotsGenerator.vb
Option Strict On

Imports AnalysisManagerBase
Imports System.IO

Public Class clsDeconToolsQCPlotsGenerator

	Protected mDebugLevel As Integer
	Protected mErrorMessage As String
	Protected mMSFileInfoScannerDLLPath As String

	Protected WithEvents mMSFileInfoScanner As MSFileInfoScannerInterfaces.iMSFileInfoScanner
	Protected mMSFileInfoScannerErrorCount As Integer

	Public ReadOnly Property ErrorMessage As String
		Get
			Return mErrorMessage
		End Get
	End Property

	Public ReadOnly Property MSFileInfoScannerErrorCount As Integer
		Get
			Return mMSFileInfoScannerErrorCount
		End Get
	End Property

    Public Sub New(MSFileInfoScannerDLLPath As String, DebugLevel As Integer)
        mMSFileInfoScannerDLLPath = MSFileInfoScannerDLLPath
        mDebugLevel = DebugLevel

        mErrorMessage = String.Empty
    End Sub

    Public Function CreateQCPlots(strInputFilePath As String, strOutputFolderPath As String) As Boolean

        Dim blnSuccess As Boolean

        Try

            mMSFileInfoScannerErrorCount = 0

            ' Initialize the MSFileScanner class					
            mMSFileInfoScanner = LoadMSFileInfoScanner(mMSFileInfoScannerDLLPath)
            With mMSFileInfoScanner
                .CheckFileIntegrity = False
                .CreateDatasetInfoFile = False
                .CreateScanStatsFile = False
                .SaveLCMS2DPlots = True
                .SaveTICAndBPIPlots = True
                .UpdateDatasetStatsTextFile = False
            End With

            blnSuccess = mMSFileInfoScanner.ProcessMSFileOrFolder(strInputFilePath, strOutputFolderPath)

            If Not blnSuccess Then
                mErrorMessage = "Error generating QC Plots using " & strInputFilePath
                Dim strMsgAddnl As String = mMSFileInfoScanner.GetErrorMessage

                If Not String.IsNullOrEmpty(strMsgAddnl) Then
                    mErrorMessage = mErrorMessage & ": " & strMsgAddnl
                End If
            End If

        Catch ex As Exception
            mErrorMessage = "Exception in CreateQCPlots: " & ex.Message
            Return False
        End Try

        Return blnSuccess

    End Function

	Protected Function LoadMSFileInfoScanner(strMSFileInfoScannerDLLPath As String) As MSFileInfoScannerInterfaces.iMSFileInfoScanner
		Const MsDataFileReaderClass As String = "MSFileInfoScanner.clsMSFileInfoScanner"

		Dim objMSFileInfoScanner As MSFileInfoScannerInterfaces.iMSFileInfoScanner = Nothing
		Dim msg As String

		Try
			If Not File.Exists(strMSFileInfoScannerDLLPath) Then
				msg = "DLL not found: " + strMSFileInfoScannerDLLPath
				clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, msg)
			Else
                Dim obj = LoadObject(MsDataFileReaderClass, strMSFileInfoScannerDLLPath)
				If obj IsNot Nothing Then
					objMSFileInfoScanner = DirectCast(obj, MSFileInfoScannerInterfaces.iMSFileInfoScanner)
					msg = "Loaded MSFileInfoScanner from " + strMSFileInfoScannerDLLPath
					If mDebugLevel >= 2 Then
						clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, msg)
					End If
				End If

			End If
		Catch ex As Exception
			msg = "Exception loading class " + MsDataFileReaderClass + ": " + ex.Message
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, msg)
		End Try

		Return objMSFileInfoScanner
	End Function

	Protected Function LoadObject(className As String, strDLLFilePath As String) As Object
		Dim obj As Object = Nothing
		Try
			' Dynamically load the specified class from strDLLFilePath
            Dim assem = Reflection.Assembly.LoadFrom(strDLLFilePath)
			Dim dllType As Type = assem.[GetType](className, False, True)
			obj = Activator.CreateInstance(dllType)
		Catch ex As Exception
			Dim msg As String = "Exception loading DLL " + strDLLFilePath + ": " + ex.Message
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, msg)
		End Try
		Return obj
	End Function

	Protected Sub mMSFileInfoScanner_ErrorEvent(Message As String) Handles mMSFileInfoScanner.ErrorEvent
		mMSFileInfoScannerErrorCount += 1
		clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "MSFileInfoScanner error: " & Message)
	End Sub

	Protected Sub mMSFileInfoScanner_MessageEvent(Message As String) Handles mMSFileInfoScanner.MessageEvent
		If mDebugLevel >= 3 Then
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, " ... " & Message)
		End If
	End Sub

End Class

File: Plugins/AM_Decon2ls_PlugIn_Decon2LSV2/clsDeconToolsQCPlotsGenerator.cs
using PRISM;
using System;
using System.IO;

namespace AnalysisManagerDecon2lsV2PlugIn
{
    public class clsDeconToolsQCPlotsGenerator : EventNotifier
    {
        private const int MAX_RUNTIME_HOURS = 5;

        private readonly int mDebugLevel;
        private readonly string mMSFileInfoScannerDLLPath;

        private MSFileInfoScannerInterfaces.iMSFileInfoScanner mMSFileInfoScanner;
        private string mInputFilePath;
        private string mOutputFolderPath;
        private bool mSuccess;

        public string ErrorMessage { get; private set; }

        public int MSFileInfoScannerErrorCount { get; private set; }

        public clsDeconToolsQCPlotsGenerator(string msFileInfoScannerDLLPath, int debugLevel)
        {
            mMSFileInfoScannerDLLPath = msFileInfoScannerDLLPath;
            mDebugLevel = debugLevel;

            ErrorMessage = string.Empty;
        }

        public bool CreateQCPlots(string inputFilePath, string outputFolderPath)
        {
            try
            {
                MSFileInfoScannerErrorCount = 0;

                // Initialize the MSFileScanner class
                mMSFileInfoScanner = LoadMSFileInfoScanner(mMSFileInfoScannerDLLPath);
                RegisterEvents(mMSFileInfoScanner);

                mMSFileInfoScanner.Options.CheckFileIntegrity = false;
                mMSFileInfoScanner.Options.CreateDatasetInfoFile = false;
                mMSFileInfoScanner.Options.CreateScanStatsFile = false;
                mMSFileInfoScanner.Options.SaveLCMS2DPlots = true;
                mMSFileInfoScanner.Options.SaveTICAndBPIPlots = true;
                mMSFileInfoScanner.Options.UpdateDatasetStatsTextFile = false;
                mMSFileInfoScanner.Options.PlotWithPython = true;

                mInputFilePath = inputFilePath;
                mOutputFolderPath = outputFolderPath;
                mSuccess = false;

                var thread = new System.Threading.Thread(ProcessMSFileOrFolderThread);

                thread.SetApartmentState(System.Threading.ApartmentState.STA);
                thread.Start();

                var startTime = DateTime.UtcNow;

                while (true)
                {
                    // Wait for 2 seconds
                    thread.Join(2000);

                    // Check whether the thread is still running
                    if (!thread.IsAlive)
                        break;

                    // Check whether the thread has been running too long
                    if (!(DateTime.UtcNow.Subtract(startTime).TotalHours > MAX_RUNTIME_HOURS))
                        continue;

                    OnErrorEvent("MSFileInfoScanner has run for over " + MAX_RUNTIME_HOURS + " hours; aborting");

                    try
                    {
                        thread.Abort();
                    }
                    catch
                    {
                        // Ignore errors here;
                    }

                    break;
                }


                if (!mSuccess)
                {
                    ErrorMessage = "Error generating QC Plots using " + inputFilePath;
                    var msgAddnl = mMSFileInfoScanner.GetErrorMessage();

                    if (!string.IsNullOrEmpty(msgAddnl))
                    {
                        ErrorMessage = ErrorMessage + ": " + msgAddnl;
                    }
                }
            }
            catch (Exception ex)
            {
                ErrorMessage = "Exception in CreateQCPlots: " + ex.Message;
                return false;
            }

            return mSuccess;
        }

        private MSFileInfoScannerInterfaces.iMSFileInfoScanner LoadMSFileInfoScanner(string msFileInfoScannerDLLPath)
        {
            const string MsDataFileReaderClass = "MSFileInfoScanner.clsMSFileInfoScanner";

            MSFileInfoScannerInterfaces.iMSFileInfoScanner msFileInfoScanner = null;
            string msg;

            try
            {
                if (!File.Exists(msFileInfoScannerDLLPath))
                {
                    msg = "DLL not found: " + msFileInfoScannerDLLPath;
                    OnErrorEvent(msg);
                }
                else
                {
                    var obj = LoadObject(MsDataFileReaderClass, msFileInfoScannerDLLPath);
                    if (obj != null)
                    {
                        msFileInfoScanner = (MSFileInfoScannerInterfaces.iMSFileInfoScanner) obj;
                        msg = "Loaded MSFileInfoScanner from " + msFileInfoScannerDLLPath;
                        if (mDebugLevel >= 1)
                        {
                            OnDebugEvent(msg);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                msg = "Exception loading class " + MsDataFileReaderClass + ": " + ex.Message;
                OnErrorEvent(msg, ex);
            }

            return msFileInfoScanner;
        }

        private object LoadObject(string className, string dllFilePath)
        {
            object obj = null;
            try
            {
                // Dynamically load the specified class from dllFilePath
                var assembly = System.Reflection.Assembly.LoadFrom(dllFilePath);
                var dllType = assembly.GetType(className, false, true);
                obj = Activator.CreateInstance(dllType);
            }
            catch (Exception ex)
            {
                var msg = "Exception loading DLL " + dllFilePath + ": " + ex.Message;
                OnErrorEvent(msg, ex);
            }
            return obj;
        }

        private void ProcessMSFileOrFolderThread()
        {
            mSuccess = mMSFileInfoScanner.ProcessMSFileOrDirectory(mInputFilePath, mOutputFolderPath);
        }
    }
}
